### DISCUSSION TITLE
Close tab causes window already closed

### DISCUSSION THREAD

**Discussion by JuGid** [php-webdriver/php-webdriver#868] [Q&A] [answered]
Hi,

I'm using php-webdriver to make some POC. For this code below, php-webdriver returns this message and I don't really understand why since the driver is the same : 
```bash
no such window: target window already closed
from unknown error: web view not found
  (Session info: chrome=89.0.4389.82)
```

The code. I use the functional test page of your GitHub repository [page_with_frame.html](https://github.com/php-webdriver/php-webdriver/blob/main/tests/functional/web/page_with_frame.html) : 

```php
putenv('WEBDRIVER_CHROME_DRIVER=' . $webdriverPath);
$driver = ChromeDriver::start();
$driver->get('file:///page_with_frame.html');
$element = $driver->findElement(WebDriverBy::id('other_iframe_content'));
$driver->switchTo()->frame($element);
$driver->findElement(WebDriverBy::id('open-new-window'))->click()
$driver->wait()->until(WebDriverExpectedCondition::numberOfWindowsToBe(2));
$driver->close();
$driver->wait()->until(WebDriverExpectedCondition::numberOfWindowsToBe(1));
//The next line is where the code returns the error
$driver->wait()->until(WebDriverExpectedCondition::titleIs('php-webdriver test page'), 'Title is timeout');
```

I assume that if `numberOfWindowsToBe` works, my `close()` use is good as I want it to close the current tab. Also, I don't know why this error comes from the `titleIs()`. 
Do you know why I get this error ?

EDIT : I tried to add a `$driver->switchTo()->defaultContent();`  before the wait titleIs, but does not work.

Thanks,

**Comment by OndraM** [comment, marked as answer]
Hi!

First of all, what do you want to achieve? You want to switch to the new window and close the original one?

If so, you must switch the context to the newly opened window (by default webdriver will still be "focused" on the original window, so that if you call `$driver->close();`, the original window is closed, but the focus was not changed to the new window (it doesn't happen automatically). Because of that, trying to read title of already closed window will lead to "target window already closed" error.

Also note - "tab" and "window" are from webdriver point of view basically the same thing.

Have a [look to documentation](https://github.com/php-webdriver/php-webdriver/wiki/Alert,-tabs,-frames,-iframes#switch-to-the-new-window).

Try this:
```php
$driver = ChromeDriver::start();
$driver->get('http://127.0.0.1:8000/page_with_frame.html');

$element = $driver->findElement(WebDriverBy::id('other_iframe_content'));
$driver->switchTo()->frame($element);
$driver->findElement(WebDriverBy::id('open-new-window'))->click(); // this opens the new window, but $driver is still switched to the original one
$driver->wait()->until(WebDriverExpectedCondition::numberOfWindowsToBe(2));
$driver->close(); // this closes the original window
$driver->wait()->until(WebDriverExpectedCondition::numberOfWindowsToBe(1));

$windowHandles = $driver->getWindowHandles(); // get array of opened windows
$driver->switchTo()->window($windowHandles[0]); // there is now obviously only one, so we can use the first array element

$driver->wait()->until(WebDriverExpectedCondition::titleIs('php-webdriver test page'), 'Title is timeout');
```

**Reply by JuGid** [reply]
Thanks a lot for this complete answer. 
I saw the documentation bu I was thinking that when a window opens, the WebDriver automatically switch to the window that's on the screen (active for the browser). But that's not the real behavior and it explains my error. (Is it mentioned in the documentation ?)

What I wanted to do is open a new window by clicking the button, then close this new one. So now, what I really have to do is clicking the button, switch to the new window, close it, switch to the original window.

Going to write it. Thanks a lot !!

**Reply by JuGid** [reply]
To people looking for a solution. This is mine.
I have a $previous_window_queue as an array (of string). Since php arrays are ordered, I can create 5 windows, add the handle to the queue, then close 5 windows, switching to each one using array_pop($previous_window_queue). 

I don't know if it's the prettiest solution, but it works for my use case.

**Reply by OndraM** [reply]
Hi, **this is not a correct way**. There is even warning [in the documentation](https://github.com/php-webdriver/php-webdriver/wiki/Alert,-tabs,-frames,-iframes#switch-to-the-new-window) :).

> getWindowHandles() does not necessarily return window in the order in which they were opened.

([As defined](https://w3c.github.io/webdriver/#get-window-handles) in WebDriver protocol).

I don't know how you get `$previous_window_queue`, but using array_pop() may not close the previous window.

**Reply by JuGid** [reply]
I set my queue with getWindowHandle , not the getWindowHandles. So when I create new windows, what I really do is : 
- Get the current handle
- Add it to the queue
- Switch to the new window 
- etc ...

For example, this is the [WindowHandler](https://github.com/JuGid/AutoMate/blob/master/src/Handler/WindowHandler.php) and this is how I use it when I [create a window](https://github.com/JuGid/AutoMate/blob/master/src/Transformer/CreateTransformer.php) and when I [close a window](https://github.com/JuGid/AutoMate/blob/master/src/Transformer/CloseTransformer.php) (not an ad) since you can only (in my code) close the 'current' window. Btw, it's a 0.X dev version.
